GVFS/GVFS.Common/Database/PlaceholdersTable.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            if (data.IsFolder)
+            {
+                if (data.IsExpandedFolder)
+                {
+                    this.AddExpandedFolder(data.Path);
+                }
+                else if (data.IsPossibleTombstoneFolder)
+                {
+                    this.AddPossibleTombstoneFolder(data.Path);
+                }
+                else
+                {
+                    this.AddPartialFolder(data.Path);
+                }
+            }
+            else
+            {
+                this.AddFile(data.Path, data.Sha);
+            }
+        }
+
+        public void AddFile(string path, string sha)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.File, Sha = sha });
+            }
+        }
+
+        public void AddPartialFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PartialFolder });
+            }
+        }
+
+        public void AddExpandedFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.ExpandedFolder });
+            }
+        }
+
+        public void AddPossibleTombstoneFolder(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PossibleTombstoneFolder });
+            }
+        }
+
+        public void Remove(string path)
+        {
+            using (IPooledConnection pooled = this.connectionPool.GetConnection())
+            using (IDbCommand command = pooled.Connection.CreateCommand())
+            {
+                Delete(command, path);
+            }
+        }
+
+        private static void Insert(IDbCommand command, PlaceholderData placeholder)
+        {
+            command.CommandText = "INSERT OR REPLACE INTO Placeholders (path, pathType, sha) VALUES (@path, @pathType, @sha);";
+            command.AddParameter("@path", DbType.String, placeholder.Path);
+            command.AddParameter("@pathType", DbType.Int32, (int)placeholder.PathType);
+
+            if (placeholder.Sha == null)

```

Comment:
Could also do...

```csharp
command.AddParameter("@sha", DbType.String, placeholder.Sha ?? DBNull.Value);
```

Reply:
Also do we ever expect `placeholder.Sha` to be an empty string?

Reply:
Because `DBNull.Value` and `placeholder.Sha` are not the same types you cannot use the `??` operator.

>do we ever expect placeholder.Sha to be an empty string?

No, it should either be `null` or a 40 character string.  I will add some validation for this.